File: leetcode/IV1.java
package test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

@SuppressWarnings("rawtypes")
class Node implements Comparable {
	String key;
	int count;
	Node(String key, int count) {
		this.key = key;
		this.count = count;
	}
	@Override
	public int compareTo(Object o) {
		Node that = (Node) o;
		int ret = this.count - that.count;
		if(0 == ret) {
			return that.key.compareTo(key);
		}
		return ret;
	}
}

public class IV1 {
	public static ArrayList<String> retFs(int numFs,
											 int topFs,
											 List<String> possibleFs,
											 int numReqs,
											 List<String> reqs) {
		ArrayList<String> ret = new ArrayList<>();
		ArrayList<String> res = new ArrayList<>();
		Map<String, Integer> hm = new HashMap<>();
		Map<String, String> caseHm = new HashMap<>();
		String[] arr = null;
		String key = null;
		PriorityQueue<Node> smallQ = new PriorityQueue<Node>();
		Node node = null;
		
		if(topFs > numReqs) topFs = numReqs;
		if(topFs == 0) return ret;
		
		for(int i=0; i<numFs; i++) {
			String str = possibleFs.get(i).toLowerCase();
			hm.put(str, 0);
			caseHm.put(str, possibleFs.get(i));
		}
		
		for(int i=0; i<numReqs; i++) {
			Set<String> hs = new HashSet<>();
			arr = reqs.get(i).split(" ");
			for(int j=0; j<arr.length; j++) {
				key = arr[j].toLowerCase();
				if(false == hm.containsKey(key)) continue;
				if(hs.add(key) == false) continue;
				hm.put(key, hm.get(key) + 1);
			}
		}
		
		for(String k: hm.keySet()) {
			node = new Node(k, hm.get(k));
			if(topFs > 0) {
				smallQ.add(node);
				topFs--;
				continue;
			}
			if(node.compareTo(smallQ.peek()) > 0 ) {
				smallQ.poll();
				smallQ.add(node);
			}
		}
		
		while(smallQ.size() != 0) ret.add(smallQ.poll().key);
		
		//Collections.reverse(ret);
		
		for(int i=ret.size()-1; i>=0; i--) res.add(caseHm.get(ret.get(i)));
		
		return res;	
	}
	
	public static void main(String[] args) { 
		int numFs = 6;
		 int topFs = 2;
		 
		 List<String> possibleFs = new ArrayList<>();
		 possibleFs.add("dell");
		 possibleFs.add("emc");
		 possibleFs.add("hover");
		 possibleFs.add("VMWare");
		 possibleFs.add("HP");
		 possibleFs.add("solar");
		 
		 int nReqs = 7;
		 
		 List<String> reqs = new ArrayList<>();
		 reqs.add("even more dell" );
		 reqs.add("emc life" );
		 reqs.add("HP" );
		 reqs.add("HP emc two IBM" );
		 reqs.add("HP please hp" );
		 reqs.add("net hover" );
		 reqs.add("solar power" );

		 List<String> ret = retFs(numFs, topFs ,possibleFs ,nReqs ,reqs);
		 System.out.println(ret);
	}
}

File: leetcode/IV1.cpp
#include "pch.h"
#include <iostream>
#include <algorithm>  
#include <string>  
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>

#include <algorithm> 

using namespace std;

class mycomparison
{
public:
	bool operator() (const pair<string, int>& a, const pair<string, int>& b) const
	{
		if (a.second == b.second) {
			return b.first > a.first;
		}
		return a.second > b.second;
	}
};

class IV1 {
public:
	static vector<string> retFs(int numFs,
		int topFs,
		vector<string> possibleFs,
		int numReqs,
		vector<string> reqs)
	{
		unordered_map<string, int> hm;
		unordered_map<string, string> caseHm;
		vector<string> arr;
		string key;
		priority_queue<pair<string, int>, vector<pair<string, int>>, mycomparison> smallQ;
		pair<string, int> node;
		mycomparison cmp;

		if (topFs > numReqs) topFs = numReqs;
		if (topFs == 0) return vector<string>();

		for (int i = 0; i < numFs; i++) {
			string str = possibleFs[i];
			transform(str.begin(), str.end(), str.begin(), tolower);

			hm[str] = 0;
			caseHm[str] = possibleFs[i];
		}

		for (int i = 0; i < numReqs; i++) {
			unordered_set<string> hs;
			splitCommon(reqs[i], " ", arr);
			for (int j = 0; j < arr.size(); j++) {
				transform(arr[j].begin(), arr[j].end(), arr[j].begin(), tolower);
				if (hm.find(arr[j]) == hm.end()) continue;
				if (hs.find(arr[j]) != hs.end()) continue;
				hs.insert(arr[j]);
				hm[arr[j]] = hm[arr[j]] + 1;
			}
		}

		for (auto it = hm.begin(); it != hm.end(); it++) {
			cout << it->first << " " << it->second << endl;
			node.first = it->first;
			node.second = it->second;;
			if (topFs > 0) {
				smallQ.push(node);
				topFs--;
				continue;
			}

			if (cmp(node, smallQ.top())) {
				smallQ.pop();
				smallQ.push(node);
			}
		}

		cout << "-------------" << endl;

		vector<string> ret(smallQ.size());
		for (int i = smallQ.size() - 1; i >= 0; i--) {
			cout << caseHm[smallQ.top().first] << " " << smallQ.top().second << endl;
			ret[i] = caseHm[smallQ.top().first];
			smallQ.pop();
		}

		return ret;
	}


	static void splitCommon(const string &str, const string &pattern, vector<string> &res)
	{
		string strs = str + pattern;
		size_t pos = strs.find(pattern);

		res.clear();
		if (str == "")
			return;

		while (pos != strs.npos)
		{
			string temp = strs.substr(0, pos);
			res.push_back(temp);
			strs = strs.substr(pos + 1, strs.size());
			pos = strs.find(pattern);
		}

		return;
	}
};


int main() {

	int numFs = 6;
	int topFs = 3;

	vector<string> possibleFs;
	possibleFs.push_back("hover");
	possibleFs.push_back("DELL");
	possibleFs.push_back("emc");

	possibleFs.push_back("VMWare");
	possibleFs.push_back("HP");
	possibleFs.push_back("solar");

	int nReqs = 7;

	vector<string> reqs;
	reqs.push_back("even more dell");
	reqs.push_back("emc life");
	reqs.push_back("HP");
	reqs.push_back("HP emc two IBM");
	reqs.push_back("HP please hp");
	reqs.push_back("net hover");
	reqs.push_back("solar power");

	vector<string> ret = IV1::retFs(numFs, topFs, possibleFs, nReqs, reqs);
	for (auto s : ret) cout << s << endl;
}


